pkg/hydrator/asset_vulnerability_hydrator.go

```diff
@@ -0,0 +1,154 @@
+package hydrator
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/asecurityteam/nexpose-vuln-hydrator/pkg/domain"
+)
+
+// AssetVulnerabilityHydrator implements the domain.Hydrator interface for hydrating
+// an asset's vulnerabilities
+type AssetVulnerabilityHydrator struct {
+	Client domain.NexposeClient
+}
+
+// HydrateVulnerabilities takes a domain.Asset and fetches its vulnerabilities to be further
+// hydrated concurrently
+func (a *AssetVulnerabilityHydrator) HydrateVulnerabilities(asset domain.Asset) (domain.AssetVulnerabilityDetails, error) {
+	// For an asset - we first get the list of vulnerabilities for a given asset ID
+	// For each vulnerability, we can fetch these concurrently, as seen on lines 35-41
+	// For each vulnerability, we can fetch the details and solutions concurrently, as seen on lines 75-102
+	// The solutions response gives a list of solution IDs, we can again fetch the solution details concurrently as seeon on lines 134-153
+
+	// 1. fetch vulnerabilities to be hydrated - this contains contains the vulnerability ID (for further hydration)
+	// as well as the results (port and protocol information of finding)
+	vulnerabilities, err := a.Client.GetAssetVulnerabilities(asset.ID)
+	if err != nil {
+		return domain.AssetVulnerabilityDetails{}, err
+	}
+
+	vulnDetailsChan := make(chan domain.VulnerabilityDetails, len(vulnerabilities))
+	errChan := make(chan error)
```

Comment:
you'll need to make this a buffered channel as well. otherwise deadlock will occur since nothing is concurrently reading from this channel

Reply:
In this case, I'm not really sure _what_ the right length would be - we know the number of vulnerabilities, but not the number of solutions that could potentially cause errors. Is it ok to use arbitrary numbers for this?